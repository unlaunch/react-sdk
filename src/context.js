import { createContext } from 'react';

const context = createContext({ flags: {}, unlaunchClient: undefined });

const {Provider,Consumer} = context;
export { Provider, Consumer };
export default context;