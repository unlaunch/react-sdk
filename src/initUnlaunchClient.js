import  {default as ULClient} from "unlaunch-js-client-lib";
/**
 * Internal function to initialize the `ULClient`.
 */
const initUnlaunchClient = async (
  apiKey,
  flagKeySet,
  identity = 'anonymous',
  attributes = null,
  options={}
) => {
  const unlaunchClient = ULClient.initialize(
    apiKey,
    flagKeySet,
    identity,
    null,
    options
  );
  return new Promise(resolve => {

    unlaunchClient.on('ready', function () {

      let rawFlags = {};

      if (flagKeySet) {
        for (const flag in flagKeySet) {
          rawFlags[flagKeySet[flag]] = unlaunchClient.variation(flagKeySet[flag]);
        }
      }
      const flags = rawFlags;
      resolve({ flags, unlaunchClient });
    });
  });
}

export default initUnlaunchClient;
