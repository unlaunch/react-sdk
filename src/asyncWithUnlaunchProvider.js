import React, { useState, FunctionComponent } from 'react';
import { Provider } from './context';
import initUnlaunchClient from './initUnlaunchClient';

/**
 * This is an async function which initializes Unlaunch JS SDK 
 * so all flags and the ldClient are ready before the app is rendered.
 * @param config - The configuration used to initialize Unlaunch JS client sdk
 */
export default async function asyncWithUnlaunchProvider(config) {
  
  const { apiKey, flagKeys, options, attributes, identity } = config;
  const { flags, unlaunchClient } = await initUnlaunchClient(apiKey, flagKeys, identity, attributes, options);
  
  const UnlaunchProvider = ({ children }) => {
    const [unlaunchData, setUnlaunchData] = useState({
      flags,
      unlaunchClient,
    });
    return <Provider value={unlaunchData}>{children}</Provider>;
  };

  return UnlaunchProvider;
}