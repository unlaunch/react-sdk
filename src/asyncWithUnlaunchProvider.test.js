
jest.mock('./initUnlaunchClient', () => jest.fn());

import React from 'react';
import { render } from '@testing-library/react';
import initUnlaunchClient from './initUnlaunchClient';
import { Consumer } from './context';
import asyncWithUnlaunchProvider from './asyncWithUnlaunchProvider';
import '@testing-library/jest-dom/extend-expect';

const apiKey = 'apikey';
const identity = 'user_123';
const App = () => <> My App </>;
const mockInitUnlaunchClient = initUnlaunchClient;
const mockFlags = { 'test-flag-1': true, 'test-flag-2': true };
let mockUnlaunchClient;

const renderWithConfig = async (config) => {
  const UnlaunchProvider = await asyncWithUnlaunchProvider(config);

  const { getByText } = render(
    <UnlaunchProvider>
      <Consumer>{value => <span>Received: {JSON.stringify(value.flags)}</span>}</Consumer>
    </UnlaunchProvider>,
  );

  return getByText(/^Received:/);
};

describe('asyncWithUnlaunchProvider', () => {
  beforeEach(() => {
    mockUnlaunchClient = {
      on: jest.fn((e, cb) => {
        cb();
      }),
    };

    mockInitUnlaunchClient.mockImplementation(() => ({
      flags: mockFlags,
      unlaunchClient: mockUnlaunchClient,
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('async provider renders app correctly', async () => {
    const UnlaunchProvider = await asyncWithUnlaunchProvider({ apiKey });
    const { container } = render(
      <UnlaunchProvider>
        <App />
      </UnlaunchProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  test('unlaunchClient is initialised correctly', async () => {
    const options = {};
    await asyncWithUnlaunchProvider({ apiKey, flagKeys:undefined, options:{}, attributes:null,identity });

    expect(mockInitUnlaunchClient).toHaveBeenCalledWith(apiKey, undefined, identity, null, {});
  });

  test('unlaunchClient is initialised correctly with target flags', async () => {
    mockInitUnlaunchClient.mockImplementation(() => ({
      flags: { "test-flag-1": true, "test-flag-2": true },
      ldClient: mockUnlaunchClient,
    }));
    const options = {};
    const flagKeys = { 'test-flag-1': false, 'test-flag-2': false };
    const receivedNode = await renderWithConfig({ apiKey, flagKeys, options, attributes:null,identity });

    expect(mockInitUnlaunchClient).toHaveBeenCalledWith(apiKey, flagKeys, identity, null, options);
    expect(receivedNode).toHaveTextContent('{"test-flag-1":true,"test-flag-2":true}');
  });
  test('internal flags state should be initialised to all flags', async () => {
    const options = {} ;
    const receivedNode = await renderWithConfig({ apiKey, identity, options });
    expect(receivedNode).toHaveTextContent('{"test-flag-1":true,"test-flag-2":true}');
  });
});