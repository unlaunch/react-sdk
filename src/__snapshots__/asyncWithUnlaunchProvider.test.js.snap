// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncWithUnlaunchProvider async provider renders app correctly 1`] = `
<div>
   My App 
</div>
`;
